### PR TITLE
feat(planner): implement lateral join

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Implement lateral join with subquery.

Part of #13558

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13685)

### CODE REVIEW

**Comment by BohuTANG** [comment]
```
Test finished, fail fast enabled, 1 out of 1805 records failed to run
0: query failed: Databend sqllogictests error: http query error: code: 1001, Text: View dependency loop detected (view: default.list_int)
[SQL] select lhs.*, rhs.*
from list_int lhs, list_int rhs
where lhs.i2 = rhs.i2 and lhs.l3 <> rhs.l3
Error: SelfError("sqllogictest failed")
order by lhs.i, rhs.i;
at tests/sqllogictests/suites/duckdb/join/test_nested_inequality.test:13
```

File: src/query/sql/src/planner/binder/table.rs [hunk 1]
```diff
@@ -466,25 +469,36 @@ impl Binder {
...
         subquery: &Query,
         alias: &Option<TableAlias>,
     ) -> Result<(SExpr, BindContext)> {
-        // For subquery, we need to use a new context to bind it.
-        let mut new_bind_context = BindContext::with_parent(Box::new(bind_context.clone()));
-        let (s_expr, mut res_bind_context) =
-            self.bind_query(&mut new_bind_context, subquery).await?;
+        // If the subquery is a lateral subquery, we need to let it see the columns
+        // from the previous queries.
+        let (result, mut result_bind_context) = if lateral {
```
Comment (xudong963) [inline src/query/sql/src/planner/binder/table.rs:478]: `if !lateral`?

File: tests/sqllogictests/suites/query/lateral.test [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+2 1
+3 2
+3 1
+
+query II
+select * from t left join lateral(select * from t1 where t1.a = t.a) t2 on t.a = t2.a
+----
+1 1
+2 2
+3 3
```
Comment (andylokandy) [inline tests/sqllogictests/suites/query/lateral.test:44]: Consider adding explain select test

File: src/query/sql/src/planner/binder/table.rs (hunk 1 above)
Comment (leiysky) [reply src/query/sql/src/planner/binder/table.rs:478]: If not lateral, we won't let it see the previous columns just as the current implementation.

File: tests/sqllogictests/suites/query/lateral.test (hunk 2 above)
Comment (leiysky) [reply tests/sqllogictests/suites/query/lateral.test:44]: Done.

**Review by xudong963 — APPROVED** [review]

File: src/query/sql/src/planner/binder/join.rs [hunk 3]
```diff
@@ -204,6 +206,29 @@ impl Binder {
             self.eq_scalars.push((left.clone(), right.clone()));
         }
 
+        let right_prop = RelExpr::with_s_expr(&right_child).derive_relational_prop()?;
+        if !right_prop.outer_columns.is_empty() {
+            // If there are outer columns in right child, then the join is a correlated lateral join
+            let mut decorrelator = SubqueryRewriter::new(self.metadata.clone());
+            right_child = decorrelator.flatten(
```
Comment (xudong963) [inline src/query/sql/src/planner/binder/join.rs:213]: https://github.com/datafuselabs/databend/actions/runs/6848848655/job/18620765947?pr=13698

CI panic

**Pull request merged**